src/unix/syscall.c

```diff
@@ -1920,12 +1920,17 @@ struct syscall {
 static struct syscall _linux_syscalls[SYS_MAX];
 struct syscall *linux_syscalls = _linux_syscalls;
 
-static context syscall_frame;
+// this can actually change to the per-cpu miscframe; if the syscall
```

Comment:
syscall is under cli, but page faults need to be handled - in both top and deferred portions (thus the bh frame). Certainly these various frames are convoluted. We need to switch contexts when we might possibly take an exception or interrupt and don't want to lose the existing saved context. Whereas linux or some other os just pushes the frame to the stack within a thread context, we shuffle these context frames around. So the syscall context allows us to e.g. take page faults without trashing the saved thread context.

The peculiar thing about the syscall context right now is that, unless we do a direct return, we continue running on the sysall context (and stack!) after we call runloop and switch to something else - just by oversight, not design. So, while the syscall top half itself will be protected by the kernlock, that lock would be released as soon as we need to block, and we can't then call runloop and continue running kernel code on the (global) syscall_frame ... so we may as well just use the per-cpu miscframe. So clearly this needs to be cleaned up before actually handling syscalls on APs...but not necessarily for this PR.

I guess this is why I made that comment about the stack-of-frames. There isn't an absolute need to have a dedicated frame for the syscall, bhqueue processing, etc...so we could just say "push frame" any time we want to protect a saved context and might take an exception and have it grab the next free per-cpu frame...the depth wouldn't be more than a few levels anyway. And the fault handlers don't necessarily need to be per-frame, do they? We could just stick the single fault handler in the cpuinfo rather than copy them to each frame.

Reply:
just so I understand..the page faults here aren't from swapped pages clearly...they are from oversubscribed allocations? missing pages from an mmaped file? it seems possible, not suggesting necessarily better, to check for this rather than take the fault if its structurally less problematic

Reply:
Yes, oversubscribed allocations. There are programs that mmap vast regions of anonymous address space, expecting that the kernel supports (anonymous) demand paging. So at some point we needed to fault in anonymous pages. (I believe it was for ruby/sinatra...mmaping something like 700+MB of address space at a time.) We don't yet support demand paging for file-backed mmaps.

We certainly could manually install these pages in the syscall top half for any given user buffer and avoid taking a fault in a deferred operation. That is, unless there are reads that also point to huge buffers, expecting that only the needed pages would be faulted in.

I guess I don't see what the fundamental problem is with allowing deferred operations to fault in pages.

Reply:
not a problem. just a choice. I'm not sure its super important if the end goal is to allow for interrupts in kernel space for other reasons. it does seem to complicate things...so if interrupts for page faults in i/o the only reason, then we could work around that.

how does this work if the user buffer (virtual, but without backing) is a direct target of a dma?

Reply:
It is admittedly a lazy approach. Plus, validation of user buffers is in order. In such a validation process, missing pages can be installed and protection violations can be caught preemptively without resorting to a fault...which of course would be absolutely necessary for the dma case as you mention (which presently isn't an issue as we don't implement any zero-copy interfaces). So I think that, once the ingredients are in order, we could avoid relying on page faults in the kernel outside of error handling.